### PR TITLE
Add filtering details to EXPLAIN and PROFILE

### PR DESCRIPTION
**Pull request by antepusic**
This PR extends the `EXPLAIN` and `PROFILE` clauses with detailed information about the `Filter` operator.

[master < Task] PR
- [x] Check, and update documentation if necessary
- [x] Provide the full content or a guide for the final git message
> Add detailed Filter operator info to EXPLAIN and PROFILE clauses

To keep docs changelog up to date, one more thing to do:
- [x] Write a release note here, including added/changed clauses
> We’ve improved query inspection using `EXPLAIN` and `PROFILE` clauses by adding information about the operands of the Filter operator.
- [x] Tag someone from docs team in the comments

### CODE REVIEW

File: src/query/plan/pretty_print.cpp [hunk 1]
```diff
@@ -180,7 +180,7 @@ bool PlanPrinter::PreVisit(query::plan::Foreach &op) {
 }
 
 bool PlanPrinter::PreVisit(query::plan::Filter &op) {
-  WithPrintLn([](auto &out) { out << "* Filter"; });
+  WithPrintLn([&](auto &out) { out << "* " << op.ToString(); });
```
Comment (as51340) [inline src/query/plan/pretty_print.cpp:183]:
```suggestion
  WithPrintLn([&op](auto &out) { out << "* " << op.ToString(); });
```
taking all as a reference is costly

File: src/query/plan/preprocess.hpp [hunk 2]
```diff
@@ -287,6 +286,8 @@ struct FilterInfo {
 /// expressions that should be generated.
 class Filters final {
  public:
+  std::vector<FilterInfo> all_filters_;
```
Comment (as51340) [inline src/query/plan/preprocess.hpp:289]: This is not such a good practice... It is better than to create a getter method than to put field as a private

File: src/query/plan/preprocess.hpp [hunk 3]
```diff
@@ -213,7 +212,7 @@ class PatternFilterVisitor : public ExpressionVisitor<void> {
 /// Stores the symbols and expression used to filter a property.
 class PropertyFilter {
  public:
-  using Bound = ScanAllByLabelPropertyRange::Bound;
+  using Bound = utils::Bound<Expression *>;
```
Comment (as51340) [inline src/query/plan/preprocess.hpp:215]: Could you explain me why is this needed?

File: src/query/plan/operator.hpp [hunk 4]
```diff
@@ -1120,6 +1124,45 @@ class Filter : public memgraph::query::plan::LogicalOperator {
...
+  static std::string SingleFilterName(const query::plan::FilterInfo &single_filter) {
+    using Type = query::plan::FilterInfo::Type;
+    if (single_filter.type == Type::Generic) {
+      return fmt::format("Generic {{{}}}", utils::IterableToString(single_filter.used_symbols, ", ",
+                                                                   [](const auto &symbol) { return symbol.name(); }));
+    } else if (single_filter.type == Type::Id) {
+      return fmt::format("id({})", single_filter.id_filter->symbol_.name());
+    } else if (single_filter.type == Type::Label) {
+      if (single_filter.expression->GetTypeInfo() != LabelsTest::kType) {
+        LOG_FATAL("This type of label filter syntax is not supported for query inspection!");
```
Comment (as51340) [inline src/query/plan/operator.hpp:1138]: Which type? I would suggest a bit better message

File: src/query/plan/operator.cpp [hunk 5]
```diff
@@ -2450,17 +2458,18 @@ Filter::FilterCursor::FilterCursor(const Filter &self, utils::MemoryResource *me
...
   // nodes and edges.
   ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
                                 storage::View::OLD, context.frame_change_collector);
-  while (input_cursor_->Pull(frame, context)) {
+  while (input_cursor_->Pull(frame, context) || UNLIKELY(context.is_profile_query)) {
     for (const auto &pattern_filter_cursor : pattern_filter_cursors_) {
       pattern_filter_cursor->Pull(frame, context);
     }
     if (EvaluateFilter(evaluator, self_.expression_)) return true;
+    if (UNLIKELY(context.is_profile_query)) return false;
```
Comment (as51340) [inline src/query/plan/operator.cpp:2472]: This logic is a bit obscure to me. You are in the while loop if context.is_profile_query is true but then at the end you return false?

File: src/query/plan/operator.hpp [hunk 6]
```diff
@@ -1120,6 +1124,45 @@ class Filter : public memgraph::query::plan::LogicalOperator {
   std::shared_ptr<memgraph::query::plan::LogicalOperator> input_;
   std::vector<std::shared_ptr<memgraph::query::plan::LogicalOperator>> pattern_filters_;
   Expression *expression_;
+  const memgraph::query::plan::Filters all_filters_;
+
+  static std::string SingleFilterName(const query::plan::FilterInfo &single_filter) {
```
Comment (as51340) [inline src/query/plan/operator.hpp:1129]: This is good but I think even better would be that this is a method of FilterInfo struct to avoid static methods, seems like a bit better design

**Review by as51340 — CHANGES_REQUESTED** [review]

**Comment by as51340** [comment]
Good job Ante, I have few ideas how to improve it and good to go

File: src/query/plan/pretty_print.cpp (hunk 1 above)
Comment (antepusic) [reply src/query/plan/pretty_print.cpp]: Done

File: src/query/plan/preprocess.hpp (hunk 3 above)
Comment (antepusic) [reply src/query/plan/preprocess.hpp:215]: Operator.hpp must include this header because it contains the definition of `Filter::all_filters_`. To prevent a circular dependency, I redefine `Bound` here without relying on operator.hpp’s `ScanAllByLabelPropertyRange`. `Bound` itself is unchanged (compare its definition in `ScanAllByLabelPropertyRange`).

File: src/query/plan/operator.cpp (hunk 5 above)
Comment (antepusic) [reply src/query/plan/operator.cpp:2472]: The `for` loop executes in two cases:
* `input_cursor_->Pull` returns `true`
* in profile queries (even if there’s no input, the query plan should contain the branch(es) that do filtering)
This line force-exits the outer `while` loop for profile queries (in their case, the `while` condition is always true).

File: src/query/plan/preprocess.hpp (hunk 2 above)
Comment (antepusic) [reply src/query/plan/preprocess.hpp:289]: I needed it because of [this line](https://github.com/memgraph/memgraph/blob/dfef16b8bfd621ef363cd81d4ffe584a7c727732/tests/unit/query_plan_operator_to_string.cpp#L235-L236). Access to `all_filters_` is already possible via the public iterators, will add a setter for it.

File: src/query/plan/operator.hpp (hunk 4 above)
Comment (antepusic) [reply src/query/plan/operator.hpp]: Fixed (NB: for label filters, an `expression` of different type than `LabelsTest` is not forbidden, but it never appears)

File: src/query/plan/operator.hpp (hunk 6 above)
Comment (antepusic) [reply src/query/plan/operator.hpp:1129]: That makes sense! I've considered it, but decided against doing so for the following reasons:
* `SingleFilterName` is used only by `Filter::ToString`
* `ExpandVariable` has an analogous (but non-static) `OperatorName` method

File: src/query/plan/operator.hpp [hunk 7]
```diff
@@ -1120,6 +1124,45 @@ class Filter : public memgraph::query::plan::LogicalOperator {
...
+    using Type = query::plan::FilterInfo::Type;
+    if (single_filter.type == Type::Generic) {
+      return fmt::format("Generic {{{}}}", utils::IterableToString(single_filter.used_symbols, ", ",
+                                                                   [](const auto &symbol) { return symbol.name(); }));
+    } else if (single_filter.type == Type::Id) {
+      return fmt::format("id({})", single_filter.id_filter->symbol_.name());
+    } else if (single_filter.type == Type::Label) {
+      if (single_filter.expression->GetTypeInfo() != LabelsTest::kType) {
+        LOG_FATAL("Label filters not using LabelsTest are not supported for query inspection!");
+      }
```
Comment (antepusic) [inline src/query/plan/operator.hpp:1139]: NB: for label filters, an expression of different type than `LabelsTest` is not forbidden, but it never appears

File: src/query/plan/operator.cpp (hunk 5 above)
Comment (as51340) [reply src/query/plan/operator.cpp:2472]: Okay but if the context.is_profile_query is true what do you need to execute, for loop of filter evaluation? Because it is a bit not clear that you want to got to while loop just to execute it once. IMO it is better to extract the part you need, not sure if it is for loop or evaluate filter, something like this:

```
if (context.is_profile_query) {
  for loop and/or
  evaluate filter
} else {
  while (input_cursor->Pull...) ...

```

**Review by as51340 — APPROVED** [review]
Adopt the change in this comment or don't it is up to you, LGTM

File: src/query/plan/preprocess.hpp [hunk 8]
```diff
@@ -302,6 +301,8 @@ class Filters final {
   auto erase(iterator first, iterator last) { return all_filters_.erase(first, last); }
   auto erase(const_iterator first, const_iterator last) { return all_filters_.erase(first, last); }
 
+  void SetFilters(std::vector<FilterInfo> all_filters) { all_filters_ = all_filters; }
```
Comment (antoniofilipovic) [inline src/query/plan/preprocess.hpp:304]: Why not pass `all_filters` by `const &`, or move them here?

Comment (antoniofilipovic) [reply src/query/plan/preprocess.hpp:304]: I think this can accept rvalue, and then you move to all_filters_:
```
void SetFilters(std::vector<FilterInfo> &&all_filters) { all_filters_ = std::move(all_filters); }
```

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
Just one comment, good job!

File: src/query/plan/preprocess.hpp (hunk 8 above)
Comment (antepusic) [reply src/query/plan/preprocess.hpp:304]: Added

**Comment by antepusic** [comment]
@vpavicic Release note in the original comment

**Review by antoniofilipovic — APPROVED** [review]
